specs/keccak256/variable-length-input.md

```diff
@@ -0,0 +1,147 @@
+# Keccak256 For Variable Length Input
+
+We describe the circuit that can process inputs of an arbitrary size.
+
+## Plain Keccak behavior
+
+Keccak256 uses a simple iterating sponge construction to handle a variable length input. The sponge absorbs 17 words(136 bytes) of inputs, applies the `Keccak-f` permutation, and then absorb next 17 words. Once all parts of the inputs are absorbed, the sponge squeezes out the 32 bytes for the output.
+
+If the last part of the input is less than 136 bytes, a padding must be applied. The padding is a multi-rate padding, it pads single bit 1 followed by minimum number of bits 0 followed by a single bit 1, such that the result is 136 bytes.
+
+## Circuit behavior
+
+### Lookup Region
+
+This region splits the input to multiple parts, each part corresponds to their `Keccak-f` permutation round.
+This is also a lookup table for the other circuits to lookup the Keccak256 input to the output.
+
+Define:
+
+We use `Round` for the cells of the following columns in a row, it represents the information in a `keccak_f` round.
+
+The `curr: Round` and `next: Round` represent the current and the next row.
+
+Columns:
+
+- `hash_id` Sequential number for grouping different hash
+- `input_len` Length for correct padding
+- `input` 136 bytes to be absorbed in this round
+- `acc_len` How many length we have absorbed
+- `acc_input` Accumulatd bytes by random linear combination (in big-endian order)
+- `output` The base-2 `state[:4]` output from this round `keccak_f`
+- `output_word_[n]` 25 columns of the output in words of this round `keccak_f`
+- `is_end_result` This flag indicates the hash output is final. The purpose is for external circuit to lookup.
+
+Selector:
+
+- `is_last_round_of_circuit` Last round of the whole circuit (not hash), to avoid `next` wrap around.
+
+| hash_id | input_len | input | acc_len | acc_input | output | is_end_result | output_word_0 | output_word_1 | output_word_2 | output_word\_... |
+| ------: | --------: | ----- | ------: | --------- | ------ | ------------: | ------------- | ------------- | ------------- | ---------------- |
+|       0 |       150 |       |       0 |           |        |             0 |               |               |               |                  |
+|       0 |       150 |       |     136 |           |        |             1 |               |               |               |                  |
```

Comment:
I'm not following this table at all. Is this related to an entire hash?? And here we have all the permutations of all of the hashes? If that's the case, Is this trying to describe the FULL keccak hash? Including N permutations per call?

Reply:
If I'm correct, then `input` can't be a single column but rather 17 word columns or something similar no?

Reply:
Sorry, can you point to the columns of the cells again? I'm not following.